File: python/atelier_bench/reports/technician_load.py
"""Technician load report."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.report_contract import ReportRow, report_row
from ..support.abstract_report import AbstractReport

if TYPE_CHECKING:
    from ..dataset import Dataset


class TechnicianLoadReport(AbstractReport):
    SLUG = "technician-load"

    def __init__(self) -> None:
        super().__init__(self.SLUG, "Technician load")

    def rows(self, data: "Dataset") -> list[ReportRow]:
        return [report_row(technician.name, technician.utilisation()) for technician in data.technicians]

File: go/atelier/reports/technician_load.go
// Package reports carries the 24 concrete reports. Each one embeds a base
// from atelier/support, declares only Rows, and registers itself in an init()
// function; Slug, Title, CacheKey and TTLSeconds arrive by promotion.
package reports

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction: this assertion is the ONLY textual
// link between the struct and the interface it satisfies.
var _ contracts.ReportContract = (*TechnicianLoadReport)(nil)

// TechnicianLoadReport reports technician load.
type TechnicianLoadReport struct {
	support.BaseReport
}

// NewTechnicianLoadReport builds the report with its slug and title.
func NewTechnicianLoadReport() *TechnicianLoadReport {
	return &TechnicianLoadReport{BaseReport: support.NewBaseReport("technician-load", "Technician load")}
}

// Rows is the only contract method this type declares.
func (r *TechnicianLoadReport) Rows(data *dataset.Dataset) []contracts.ReportRow {
	rows := make([]contracts.ReportRow, 0, len(data.Technicians))
	for _, technician := range data.Technicians {
		rows = append(rows, contracts.Row(technician.Name, technician.Utilisation()))
	}
	return rows
}

// init registers this report; the package-level registry is built by the 24
// init functions in this package, in file-name order.
func init() {
	Register(NewTechnicianLoadReport())
}
